pipelines/commands/status.go

```diff
@@ -0,0 +1,193 @@
+	return sc
+}
+
+func (sc *StatusCommand) SetPipeline(pl string) *StatusCommand {
+	sc.pipelineName = pl
+	return sc
+}
+
+func (sc *StatusCommand) SetNotify(nf bool) *StatusCommand {
+	sc.notify = nf
+	return sc
+}
+
+func (sc *StatusCommand) SetMultiBranch(multiBranch bool) *StatusCommand {
+	sc.isMultiBranch = multiBranch
+	return sc
+}
+
+func (sc *StatusCommand) Run() (string, error) {
+	// create service manager to fetch run status
+	serviceManager, svcMgrErr := manager.CreateServiceManager(sc.serverDetails)
+	if svcMgrErr != nil {
+		return "", svcMgrErr
+	}
+
+	// get pipeline status using branch name, pipelines name and whether it is multi branch
+	matchingPipes, pipStatusErr := serviceManager.GetPipelineRunStatusByBranch(sc.branch, sc.pipelineName, sc.isMultiBranch)
+	if pipStatusErr != nil {
+		return "", errorutils.CheckError(pipStatusErr)
+	}
+	var res string
+	for i := range matchingPipes.Pipelines {
+		p := matchingPipes.Pipelines[i]
+		if p.LatestRunID != 0 { // filter out pipelines which are not run at all
+			if sc.pipelineName != "" && sc.notify { // when notification option is selected use this flow to notify
+				monErr := monitorStatusAndNotify(context.Background(), serviceManager, sc.branch, sc.pipelineName, sc.isMultiBranch)
+				if monErr != nil {
+					return "", errorutils.CheckError(monErr)
+				}
+			} else {
+				reStatus, colorCode, d := getPipelineStatusAndColorCode(&p)
+				res := colorCode.Sprintf("\n%s %s\n%14s %s\n%14s %d \n%14s %s \n%14s %s\n", "PipelineName :", p.Name, "Branch :", p.PipelineSourceBranch, "Run :", p.Run.RunNumber, "Duration :", d, "Status :", reStatus)
+				return res, nil
+			}
+		}
+	}
+	return res, nil
+}
+
+// getPipelineStatusAndColorCode based on pipeline status code
+// return color to be used for pretty printing
+func getPipelineStatusAndColorCode(pipeline *services.Pipelines) (string, color.Color, string) {
+	s := status.GetPipelineStatus(pipeline.Run.StatusCode)
+	colorCode := status.GetStatusColorCode(s)
+	d := pipeline.Run.DurationSeconds
+	if d == 0 {
+		// calculate duration it took till now by differentiating created time with current time
+		t1 := pipeline.Run.CreatedAt
+		t2 := time.Now()
+		d = int(t2.Unix() - t1.Unix())
+	}
+
+	return s, colorCode, convertSecToDay(d)
+}
+
+// ConvertSecToDay converts seconds passed as integer to
+// - D H M S format
+func convertSecToDay(sec int) string {
+	log.Debug("received duration is: ", sec)
+	day := sec / (24 * 3600)
+
+	sec = sec % (24 * 3600)
+	hour := sec / 3600
+
+	sec %= 3600
+	minutes := sec / 60
+
+	sec %= 60
+	seconds := sec
+
+	return fmt.Sprintf("%dD %dH %dM %dS", day, hour, minutes, seconds)
+}
+
+// monitorStatusAndNotify monitor for status change and
+// send notification if there is a change identified
+func monitorStatusAndNotify(ctx context.Context, pipelinesMgr *pipelines.PipelinesServicesManager, branch string, pipName string, isMultiBranch bool) error {
+	ctx, cancel := context.WithTimeout(ctx, 60*time.Minute)
```

Comment:
Did you check it using ctrl+c? Does it freeze?

Reply:
This is for long monitoring until a pipeline run is available